### PR TITLE
Basic mux tests (with check that connection does not disconnect for basic scenarios)

### PR DESCRIPTION
**Pull request by abhipsaMisra**
Basic Scenarios:
- [x] Message Send E2E muxed
- [x] Message Receive E2E muxed
- [x] Methods E2E muxed
- [x] Twin E2E muxed 
- [x] Device client - all operations over muxed connection

There is a connection status check enabled to verify the muxing does not cause previously added devices to disconnect.

Open items:
- [ ] Message, method and twin common code is still in util classes.
- [ ] Modules operations - basic and combined operations.

### CODE REVIEW

File: e2e/test/MessageReceiveE2ETests.cs [hunk 1]
```diff
@@ -113,66 +113,6 @@ private Client.Message ComposeD2CTestMessage(out string payload, out string p1Va
...
-            _log.WriteLine($"{nameof(ComposeC2DTestMessage)}: payload='{payload}' messageId='{messageId}' p1Value='{p1Value}'");
-
-            return new Message(Encoding.UTF8.GetBytes(payload))
-            {
-                MessageId = messageId,
-                Properties = { ["property1"] = p1Value }
-            };
-        }
-
-        public static async Task VerifyReceivedC2DMessageAsync(Client.TransportType transport, DeviceClient dc, string payload, string p1Value)
```
Comment (CIPop) [inline e2e/test/MessageReceiveE2ETests.cs]: This code must be near the test for clarity.

File: e2e/test/MessageSend.cs [hunk 2]
```diff
@@ -5,12 +5,14 @@
...
 using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
 using System.Text;
 using System.Threading.Tasks;
 
 namespace Microsoft.Azure.Devices.E2ETests
 {
-    public static class MessageSendUtil
+    public static class MessageSend
```
Comment (CIPop) [inline e2e/test/MessageSend.cs]: Same as above.

File: e2e/test/MessageSendE2EMultiplexingTests.cs [hunk 3]
```diff
@@ -13,18 +13,16 @@ namespace Microsoft.Azure.Devices.E2ETests
 {
     [TestClass]
     [TestCategory("IoTHub-E2E")]
-    [TestCategory("ConnectionPooling_AMQP_E2E")]
```
Comment (CIPop) [inline e2e/test/MessageSendE2EMultiplexingTests.cs]: 1. Why do we need another category.
2. Why is it different from the class name?

File: e2e/test/MessageSendE2EMultiplexingTests.cs [hunk 4]
```diff
@@ -13,18 +13,16 @@ namespace Microsoft.Azure.Devices.E2ETests
...
     public class MessageSendE2EMultiplexingTests : IDisposable
     {
         private readonly string DevicePrefix = $"E2E_{nameof(MessageSendE2EMultiplexingTests)}_";
+        private readonly string ModulePrefix = $"E2E_{nameof(MessageSendE2EMultiplexingTests)}_";
```
Comment (CIPop) [inline e2e/test/MessageSendE2EMultiplexingTests.cs]: Class name must include AMQP.

File: e2e/test/MessageSendE2EMultiplexingTests.cs [hunk 5]
```diff
@@ -13,18 +13,16 @@ namespace Microsoft.Azure.Devices.E2ETests
...
         private readonly int MuxWithoutPoolingDevicesCount = 2;
-        // For enabling multiplexing without pooling, the pool size needs to be set to 1
-        private readonly int MuxWithoutPoolingPoolSize = 1;
-        // These values are configurable
+        private readonly int MuxWithoutPoolingPoolSize = 1; // For enabling multiplexing without pooling, the pool size needs to be set to 1
         private readonly int MuxWithPoolingDevicesCount = 4;
         private readonly int MuxWithPoolingPoolSize = 2;
 
-        // TODO: Implement proxy and mux tests
+        // TODO: #839 - Implement proxy and mux tests
```
Comment (CIPop) [inline e2e/test/MessageSendE2EMultiplexingTests.cs]: Same as above: why would we add proxy?

File: e2e/test/MessageSendE2EMultiplexingTests.cs [hunk 6]
```diff
@@ -13,18 +13,16 @@ namespace Microsoft.Azure.Devices.E2ETests
...
     [TestClass]
     [TestCategory("IoTHub-E2E")]
-    [TestCategory("ConnectionPooling_AMQP_E2E")]
     public class MessageSendE2EMultiplexingTests : IDisposable
     {
         private readonly string DevicePrefix = $"E2E_{nameof(MessageSendE2EMultiplexingTests)}_";
+        private readonly string ModulePrefix = $"E2E_{nameof(MessageSendE2EMultiplexingTests)}_";
         private readonly int MuxWithoutPoolingDevicesCount = 2;
-        // For enabling multiplexing without pooling, the pool size needs to be set to 1
-        private readonly int MuxWithoutPoolingPoolSize = 1;
```
Comment (CIPop) [inline e2e/test/MessageSendE2EMultiplexingTests.cs]: Mux w/o pooling isn't possible.

File: e2e/test/MessageSendE2EMultiplexingTests.cs [hunk 7]
```diff
@@ -242,12 +236,84 @@ public async Task Message_X509_DeviceSendSingleMessage_MuxWithPooling_AmqpWs()
                 foreach (DeviceClient deviceClient in deviceClients)
                 {
                     await deviceClient.CloseAsync().ConfigureAwait(false);
-                    _log.WriteLine($"{nameof(MessageSendE2ETests)}: Disposing deviceClient {TestLogging.GetHashCode(deviceClient)}");
```
Comment (CIPop) [inline e2e/test/MessageSendE2EMultiplexingTests.cs]: Operations are performed in the `finally` block.

File: e2e/test/MessageSendE2EMultiplexingTests.cs [hunk 8]
```diff
@@ -242,12 +236,84 @@ public async Task Message_X509_DeviceSendSingleMessage_MuxWithPooling_AmqpWs()
...
+
+                    // The connection status change count should be 2: connect (open) and disabled (close)
+                    Assert.IsTrue(deviceClientConnectionStatusChangeCount[deviceClient] == 2);
```
Comment (CIPop) [inline e2e/test/MessageSendE2EMultiplexingTests.cs]: Please use AAA.

File: e2e/test/MessageSendE2EMultiplexingTests.cs [hunk 9]
```diff
@@ -242,12 +236,84 @@ public async Task Message_X509_DeviceSendSingleMessage_MuxWithPooling_AmqpWs()
...
+                    string devicePrefix = useSameDevice ? DevicePrefix : $"{DevicePrefix}_{i}_";
+                    string modulePrefix = useSameDevice ? $"{ModulePrefix}_{i}_" : ModulePrefix;
+
+                    TestModule testModule = await TestModule.GetTestModuleAsync(devicePrefix, modulePrefix).ConfigureAwait(false);
+                    ModuleClient moduleClient = ModuleClient.CreateFromConnectionString(testModule.ConnectionString, transportSettings);
+                    moduleClients.Add(moduleClient);
+
+                    moduleClient.SetConnectionStatusChangesHandler((status, statusChangeReason) =>
+                    {
+                        setConnectionStatusChangesHandlerCount++;
```
Comment (CIPop) [inline e2e/test/MessageSendE2EMultiplexingTests.cs]: Same as above - if at all possible, avoid using lambdas with captures where lifetime of the variables is not clear.

File: e2e/test/MessageSendE2EMultiplexingTests.cs [hunk 10]
```diff
@@ -242,12 +236,84 @@ public async Task Message_X509_DeviceSendSingleMessage_MuxWithPooling_AmqpWs()
...
+
+                    _log.WriteLine($"{nameof(MessageSendE2EMultiplexingTests)}: Preparing to send message for module {i}");
+                    await moduleClient.OpenAsync().ConfigureAwait(false);
+                    await MessageSend.SendSingleMessageModuleAndVerifyAsync(moduleClient, testModule.DeviceId).ConfigureAwait(false);
+                }
+            }
+            finally
+            {
+                // Close and dispose all of the module client instances here
+                foreach (ModuleClient moduleClient in moduleClients)
```
Comment (CIPop) [inline e2e/test/MessageSendE2EMultiplexingTests.cs]: No non-cleanup actions in the `finally` block. (e.g. calling `CloseAsync` or `Assert`).
Use `using` for dispose as much as possible instead of implementing your own.

File: e2e/test/MessageSendFaultInjectionTests.cs [hunk 11]
```diff
@@ -146,78 +137,119 @@ public async Task Message_AmqpD2CLinkDropSendRecovery_AmqpWs()
...
                     TestDeviceType.Sasl,
                     Client.TransportType.Amqp_WebSocket_Only,
                     FaultInjection.FaultType_Throttle,
                     FaultInjection.FaultCloseReason_Boom,
                     FaultInjection.DefaultDelayInSec,
                     FaultInjection.DefaultDurationInSec).ConfigureAwait(false);
+            }
+            catch (IotHubException ex)
+            {
+                Assert.IsInstanceOfType(ex, typeof(IotHubThrottledException));
```
Comment (CIPop) [inline e2e/test/MessageSendFaultInjectionTests.cs]: Catch the exception that you expect.

File: e2e/test/MessageSendFaultInjectionTests.cs [hunk 12]
```diff
@@ -146,78 +137,119 @@ public async Task Message_AmqpD2CLinkDropSendRecovery_AmqpWs()
...
-
-        // WIP: Exception handling
+        
         [TestMethod]
         public async Task Message_ThrottledConnectionRecovery_Amqp()
         {
-            await SendMessageRecovery(
+            try
+            {
+                await SendMessageRecovery(
```
Comment (CIPop) [inline e2e/test/MessageSendFaultInjectionTests.cs]: Why would we expect an error on a recovery test that, from the name, it's supposed to pass.
Please name the tests like _ComponentUnderTest___Action___ExpectedResult__[_Protocol_]

File: e2e/test/MessageSendFaultInjectionTests.cs [hunk 13]
```diff
@@ -146,78 +137,119 @@ public async Task Message_AmqpD2CLinkDropSendRecovery_AmqpWs()
...
                     TestDeviceType.Sasl,
                     Client.TransportType.Amqp_Tcp_Only,
                     FaultInjection.FaultType_Throttle,
                     FaultInjection.FaultCloseReason_Boom,
                     FaultInjection.DefaultDelayInSec,
                     FaultInjection.DefaultDurationInSec).ConfigureAwait(false);
+            }
+            catch (IotHubException ex)
+            {
+                Assert.IsInstanceOfType(ex, typeof(IotHubThrottledException));
```
Comment (CIPop) [inline e2e/test/MessageSendFaultInjectionTests.cs]: Catch the right exception?

File: e2e/test/MethodE2EMultiplexingTests.cs [hunk 14]
```diff
@@ -13,17 +13,16 @@ namespace Microsoft.Azure.Devices.E2ETests
...
     [TestCategory("IoTHub-E2E")]
-    [TestCategory("ConnectionPooling_AMQP_E2E")]
     public class MethodE2EMultiplexingTests : IDisposable
     {
         private readonly string DevicePrefix = $"E2E_{nameof(MethodE2EMultiplexingTests)}_";
         private readonly int MuxWithoutPoolingDevicesCount = 2;
-        // For enabling multiplexing without pooling, the pool size needs to be set to 1
-        private readonly int MuxWithoutPoolingPoolSize = 1;
-        // These values are configurable
+        private readonly int MuxWithoutPoolingPoolSize = 1; // For enabling multiplexing without pooling, the pool size needs to be set to 1
```
Comment (CIPop) [inline e2e/test/MethodE2EMultiplexingTests.cs]: Mux w/o pooling isn't possible. The comment is wrong.

File: e2e/test/MethodE2EMultiplexingTests.cs [hunk 15]
```diff
@@ -13,17 +13,16 @@ namespace Microsoft.Azure.Devices.E2ETests
...
         private readonly int MuxWithPoolingDevicesCount = 4;
         private readonly int MuxWithPoolingPoolSize = 2;
 
+        // TODO: #839 - Implement proxy and mux tests
+        private static string ProxyServerAddress = Configuration.IoTHub.ProxyServerAddress;
```
Comment (CIPop) [inline e2e/test/MethodE2EMultiplexingTests.cs]: Same as above: why is this needed?

File: e2e/test/MethodE2EMultiplexingTests.cs [hunk 16]
```diff
@@ -158,18 +244,32 @@ public async Task Method_IoTHubSak_DeviceReceivesMethodAndResponse_MuxWithPoolin
...
+                    ConnectionStatusChangeReason? lastConnectionStatusChangeReason = null;
+                    int setConnectionStatusChangesHandlerCount = 0;
+
+                    TestDevice testDevice = await TestDevice.GetTestDeviceAsync($"{DevicePrefix}_{i}_").ConfigureAwait(false);
                     DeviceClient deviceClient = testDevice.CreateDeviceClient(transportSettings, authScope);
                     deviceClients.Add(deviceClient);
 
+                    deviceClient.SetConnectionStatusChangesHandler((status, statusChangeReason) =>
+                    {
+                        setConnectionStatusChangesHandlerCount++;
```
Comment (CIPop) [inline e2e/test/MethodE2EMultiplexingTests.cs]: Same comments from the other tests.
Looks like this code is common: please extract the common pattern (use template method for example) and apply to all tests. (For an example see the FaultInjection tests.)

File: common/test/TestDevice.cs [hunk 17]
```diff
@@ -195,31 +206,5 @@ private static string GetHostName(string iotHubConnectionString)
             Regex regex = new Regex("HostName=([^;]+)", RegexOptions.None);
             return regex.Match(iotHubConnectionString).Groups[1].Value;
         }
-
-        public static ITransportSettings[] GetTransportSettings(Client.TransportType transportType)
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply common/test/TestDevice.cs]: This has been removed since we are not converting transport type into transport settings in the test class anymore.

File: e2e/test/CombinedClientOperationsMultiplexingTests.cs [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.E2ETests
+{
+    [TestClass]
+    [TestCategory("IoTHub-E2E")]
+    public class CombinedClientOperationsMultiplexingTests : IDisposable
+    {
+        private readonly string DevicePrefix = $"E2E_{nameof(CombinedClientOperationsMultiplexingTests)}_";
+        private readonly int MuxWithoutPoolingDevicesCount = 2;
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply e2e/test/CombinedClientOperationsMultiplexingTests.cs]: Multiplexing without pooling is the scenario when we have 1 connection and multiple devices share the same connection.

File: e2e/test/FaultInjection.cs [hunk 19]
```diff
@@ -125,153 +124,22 @@ public static async Task ActivateFaultInjection(Client.TransportType transport,
...
-                await deviceClient.CloseAsync().ConfigureAwait(false);
-                await cleanupOperation().ConfigureAwait(false);
-
-                s_log.WriteLine($"{nameof(FaultInjection)}: Disposing deviceClient {TestLogging.GetHashCode(deviceClient)}");
-                deviceClient.Dispose();
-            }
-        }
-
-        // Error injection template method (multiplexing over amqp).
-        public static async Task TestErrorInjectionMuxedOverAmqpAsync(
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply e2e/test/FaultInjection.cs]: This is a revert; these methods are now removed.

File: e2e/test/FaultInjection.cs [hunk 20]
```diff
@@ -312,33 +179,36 @@ public static async Task ActivateFaultInjection(Client.TransportType transport,
                 s_log.WriteLine($"{nameof(FaultInjection)}: Waiting for fault injection to be active: {delay}ms");
                 await Task.Delay(delay).ConfigureAwait(false);
 
-                // TODO: This needs to be changed - for recoverable faults we need to wait some more time for the client to know about fault and then retry
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply e2e/test/FaultInjection.cs]: This is a revert; these comments are now removed.

File: e2e/test/MessageSendE2EMultiplexingTests.cs (hunk 3 above)
Comment (abhipsaMisra) [reply e2e/test/MessageSendE2EMultiplexingTests.cs]: This is a revert; these categories are now removed.

File: common/test/TestDevice.cs [hunk 21]
```diff
@@ -157,11 +157,22 @@ public Client.IAuthenticationMethod AuthenticationMethod
...
+        public DeviceClient CreateDeviceClient(Client.TransportType transport)
         {
-            ITransportSettings[] transportSettings = GetTransportSettings(transport);
-            return CreateDeviceClient(transportSettings, authScope);
+            DeviceClient deviceClient = null;
+
+            if (_authenticationMethod == null)
+            {
+                deviceClient = DeviceClient.CreateFromConnectionString(ConnectionString, transport);
+                s_log.WriteLine($"{nameof(CreateDeviceClient)}: Created {nameof(DeviceClient)} from connection string: {transport} ID={TestLogging.IdOf(deviceClient)}");
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply common/test/TestDevice.cs]: Logging the device name with device ID

**Comment by CIPop** [comment]
Tracking PR comments with https://github.com/Azure/azure-iot-sdk-csharp/issues/889.
